### PR TITLE
[d3d9] Fix various issues in UpdateTexture

### LINKED ISSUE #1947: GTR2 rendering issue d3d9
Please describe your issue as accurately as possible. If you run into a problem with a binary release, make sure to test with latest `master` as well.

With d3d9 after commit 338f6dfb0efe6536c5954ae0d6fba60809952da8 I see rendering issues, black areas in first screen grab, the second one is there for comparison.

Grab_001
Grab_003

### Software information
Name of the game, settings used etc.
GTR2 d3d9
### System information
- GPU: nvidia 780ti
- Driver: nvidia binary 460.56
- Wine version: 6.2 staging
- DXVK version:  latest git db69ade7e7a7d460fef8893d8bb5c1922fa8ba64
- and a0cf5926d89a812b9932e793e34e036fe68c6230

### Apitrace file(s)
to follow if required

### Log files
- d3d9.log: 
[GTR2_d3d9.log](https://github.com/doitsujin/dxvk/files/6050711/GTR2_d3d9.log)
[GTR2_d3d9-working.log](https://github.com/doitsujin/dxvk/files/6050748/GTR2_d3d9-working.log)

- d3d11.log:
- dxgi.log:

### PR DESCRIPTION
**Pull request by K0bin**
As discussed on Discord.

Closes #1947

### CODE REVIEW

**Comment by Joshua-Ashton** [comment]
Can you make it very clear what extents/offsets come from either the box or from the size of the actual image's mip.

It'd make this code so much easier to follow.

File: src/d3d9/d3d9_device.cpp [hunk 1]
```diff
@@ -754,40 +754,41 @@ namespace dxvk {
...
-                                  + srcByteOffset.x * formatInfo->elementSize;
-        VkExtent2D srcExtent = VkExtent2D{ blockCount.width  * formatInfo->blockSize.width,
-                                           blockCount.height * formatInfo->blockSize.height };
+        VkExtent3D scaledBoxExtent = util::computeMipLevelExtent({
+          uint32_t(box.Right - box.Left),
+          uint32_t(box.Bottom - box.Top),
+          uint32_t(box.Back - box.Front)
+        }, m);
+        VkExtent3D scaledBoxBlockCount = util::computeBlockCount(scaledBoxExtent, formatInfo->blockSize);
+        VkExtent3D scaledAlignedBoxExtend = util::computeBlockExtent(scaledBoxBlockCount, formatInfo->blockSize);
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:775]: *extent

File: src/d3d9/d3d9_device.cpp [hunk 2]
```diff
@@ -754,40 +754,41 @@ namespace dxvk {
...
-          alignDown(uint32_t((box.Back - box.Front) >> m), formatInfo->blockSize.depth),
-        };
-
-        VkExtent3D levelExtent = dstImage->mipLevelExtent(m);
-        VkExtent3D blockCount  = util::computeBlockCount(levelExtent, formatInfo->blockSize);
-        VkOffset3D srcByteOffset = {
-          offset.x / int32_t(formatInfo->blockSize.width),
-          offset.y / int32_t(formatInfo->blockSize.height),
-          offset.z / int32_t(formatInfo->blockSize.depth)
+        VkOffset3D boxOffset = {
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:764]: scaledBoxOffset

File: src/d3d9/d3d9_device.cpp [hunk 3]
```diff
@@ -754,40 +754,41 @@ namespace dxvk {
...
+        VkExtent3D scaledBoxExtent = util::computeMipLevelExtent({
+          uint32_t(box.Right - box.Left),
+          uint32_t(box.Bottom - box.Top),
+          uint32_t(box.Back - box.Front)
+        }, m);
+        VkExtent3D scaledBoxBlockCount = util::computeBlockCount(scaledBoxExtent, formatInfo->blockSize);
+        VkExtent3D scaledAlignedBoxExtend = util::computeBlockExtent(scaledBoxBlockCount, formatInfo->blockSize);
+
+        VkExtent3D texLevelExtent = dstImage->mipLevelExtent(m);
+        VkExtent3D texBlockCount  = util::computeBlockCount(texLevelExtent, formatInfo->blockSize);
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:778]: It's long but, texLevelExtentBlockCount

File: src/d3d9/d3d9_device.cpp [hunk 4]
```diff
@@ -754,40 +754,41 @@ namespace dxvk {
...
-                                  + srcByteOffset.y * formatInfo->elementSize * blockCount.width
-                                  + srcByteOffset.x * formatInfo->elementSize;
-        VkExtent2D srcExtent = VkExtent2D{ blockCount.width  * formatInfo->blockSize.width,
-                                           blockCount.height * formatInfo->blockSize.height };
+        VkExtent3D scaledBoxExtent = util::computeMipLevelExtent({
+          uint32_t(box.Right - box.Left),
+          uint32_t(box.Bottom - box.Top),
+          uint32_t(box.Back - box.Front)
+        }, m);
+        VkExtent3D scaledBoxBlockCount = util::computeBlockCount(scaledBoxExtent, formatInfo->blockSize);
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:774]: scaledBoxExtentBlockCount

File: src/d3d9/d3d9_device.cpp [hunk 5]
```diff
@@ -754,40 +754,41 @@ namespace dxvk {
...
+        VkExtent3D scaledAlignedBoxExtend = util::computeBlockExtent(scaledBoxBlockCount, formatInfo->blockSize);
+
+        VkExtent3D texLevelExtent = dstImage->mipLevelExtent(m);
+        VkExtent3D texBlockCount  = util::computeBlockCount(texLevelExtent, formatInfo->blockSize);
+        VkOffset3D srcBlockOffset = util::computeBlockOffset(boxOffset, formatInfo->blockSize);
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:779]: boxOffsetBlockCount

File: src/d3d9/d3d9_device.cpp (hunk 3 above)
Comment (Joshua-Ashton) [reply src/d3d9/d3d9_device.cpp:778]: I'd also like us to change this to computeBlockExtent for clarity but we can do that later...

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_device.cpp (hunk 1 above)
Comment (K0bin) [reply src/d3d9/d3d9_device.cpp:775]: oops

**Pull request merged**